File: 2022/src/bin/e01.hs
module Main where

import Data.List.Split
import Data.Ord
import Text.Read
import Data.Maybe
import Data.List
import Control.Monad

main :: IO ()
main = do
  run a "s01" 24000
  run a "e01" 69795
  run b "s01" 45000
  run b "e01" 208437

run :: (Monad m, Eq a, Show a) => (String -> m a) -> String -> a -> m ()
run f g expected = do
  actual <- f $ "txt/" ++ g ++ ".txt"
  when (expected /= actual) $
    error $ concat [g, "expected: ", show expected, ", actual: ", show actual]

a :: FilePath -> IO Int
a = sol 1

b :: FilePath -> IO Int
b = sol 3

sol :: Int -> String -> IO Int
sol n fp
  = sum . take n
  . sortBy (flip (comparing id))
  . map (sum . map (fromMaybe 0))
  . splitOn [Nothing]
  . map readMaybe
  . lines <$> readFile fp

File: 2022/src/bin/e01.rs
use aoc2021::*;

fn main() {
    run()
}

#[test]
fn test() {
    run()
}

fn run() {
    assert_eq!(a("txt/s01.txt"), 24000);
    assert_eq!(a("txt/e01.txt"), 69795);
    assert_eq!(b("txt/s01.txt"), 45000);
    assert_eq!(b("txt/e01.txt"), 208437);
}

fn a(s: &str) -> i64 {
    sol(s)[0]
}

fn b(s: &str) -> i64 {
    sol(s).iter().take(3).sum()
}

fn sol(s: &str) -> Vec<i64> {
    let lines: Vec<String> = read_parsed(s).collect();
    let mut elves = vec![];
    let mut this_elf = 0;

    for line in lines {
        if line == "" {
            elves.push(this_elf);
            this_elf = 0;
        } else {
            let kcal: i64 = line.parse().unwrap();
            this_elf += kcal;
        }
    }
    if this_elf != 0 {
        elves.push(this_elf);
    }

    elves.sort_by(|a, b| b.cmp(a));
    elves
}
